annotation/tests/test_tasks_services.py

```diff
@@ -295,3 +338,983 @@ def test_validate_user_actions_invalid_states(
+            [call(mock_session, user_id) for user_id in expected_user_ids],
+            any_order=True,
+        )
+
+
+def test_create_tasks_with_empty_tasks(mock_session: Mock):
+    tasks = []
+    job_id = 1
+    with patch(
+        "annotation.tasks.services.update_files"
+    ) as mock_update_files, patch(
+        "annotation.tasks.services.update_user_overall_load"
+    ) as mock_update_user_overall_load:
+        mock_bulk_insert_mappings = MagicMock()
+        mock_session.bulk_insert_mappings = mock_bulk_insert_mappings
+        mock_session.flush = MagicMock()
+
+        create_tasks(mock_session, tasks, job_id)
+
+        assert mock_bulk_insert_mappings.call_count == 1
+        assert mock_update_files.call_count == 1
+        mock_update_user_overall_load.assert_not_called()
+
+
+def create_task(status: TaskStatusEnumSchema) -> ManualAnnotationTask:
+    task = MagicMock(spec=ManualAnnotationTask)
+    task.status = status
+    task.id = 1
+    return task
+
+
+def test_update_task_status_ready(mock_session: Mock):
+    task = create_task(TaskStatusEnumSchema.ready)
+
+    update_task_status(mock_session, task)
+
+    assert task.status == TaskStatusEnumSchema.in_progress
+    mock_session.add.assert_called_once_with(task)
+    mock_session.commit.assert_called_once()
+
+
+@pytest.mark.parametrize(
+    ("status", "expected_message"),
+    (
+        (TaskStatusEnumSchema.pending, "Job is not started yet"),
+        (TaskStatusEnumSchema.finished, "Task is already finished"),
+    ),
+)
+def test_update_task_status_error(
+    mock_session: Mock, status: TaskStatusEnumSchema, expected_message: str
+):
+    task = create_task(status)
+    with pytest.raises(FieldConstraintError, match=f".*{expected_message}.*"):
+        update_task_status(mock_session, task)
+
+
+def test_finish_validation_task(mock_session: MagicMock):
+    mock_task = create_task(TaskStatusEnumSchema.ready)
+    mock_query = MagicMock()
+    mock_session.query.return_value = mock_query
+    mock_query.filter.return_value = mock_query
+    mock_query.with_for_update.return_value = mock_query
+    mock_query.update.return_value = None
+
+    finish_validation_task(mock_session, mock_task)
+
+    mock_session.query.assert_called_once_with(ManualAnnotationTask)
+
+    mock_query.with_for_update.assert_called_once()
+    mock_query.update.assert_called_once_with(
+        {ManualAnnotationTask.status: TaskStatusEnumSchema.finished},
+        synchronize_session="fetch",
+    )
+    mock_session.commit.assert_called_once()
+
+
+def test_count_annotation_tasks(mock_session: Mock):
+    mock_task = create_task(TaskStatusEnumSchema.ready)
+    mock_query = MagicMock()
+    mock_session.query.return_value = mock_query
+    mock_query.filter.return_value = mock_query
+    mock_query.count.return_value = 5
+
+    result = count_annotation_tasks(mock_session, mock_task)
+
+    mock_session.query.assert_called_once_with(ManualAnnotationTask)
```

Comment:
Can we add an assertion test for the filter method here?
```suggestion
    mock_session.query.assert_called_once_with(ManualAnnotationTask)
        mock_session.filter.assert_called_once_with(
            ManualAnnotationTask.job_id == mock_task.job_id,
            ManualAnnotationTask.file_id == mock_task.file_id,
            ManualAnnotationTask.is_validation.is_(False),
    )
```